### PR TITLE
[SPARK-42337][SQL] Add error class INVALID_TEMP_OBJ_REFERENCE

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?

This PR adds a new error class `INVALID_TEMP_OBJ_REFERENCE ` and replaces two existing error classes with this new one:
- _LEGACY_ERROR_TEMP_1283
- _LEGACY_ERROR_TEMP_1284

### Why are the changes needed?

To improve the error messages.

### Does this PR introduce _any_ user-facing change?

Yes, the PR changes a user-facing error message.

### How was this patch tested?

Existing unit tests.

### CODE REVIEW

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -177,6 +177,11 @@
     ],
     "sqlState" : "22018"
   },
+  "CREATE_PERSISTENT_OBJECT_OVER_TEMP_OBJECT" : {
+    "message" : [
+      "Cannot create a persistent <obj> <objName> by referencing a temporary <tempObj> <tempObjName>. Please make the temporary <tempObj> persistent, or make the persistent <obj> temporary."
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:182]: Is there any difference among `persistent` and `permanent`? The original errors talk about `permanent` but your named the object differently. Just wonder why?

File: core/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -177,6 +177,11 @@
     ],
     "sqlState" : "22018"
   },
+  "CREATE_PERSISTENT_OBJECT_OVER_TEMP_OBJECT" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:180]: Could you make it shorter? like
```suggestion
  "CANNOT_REFER_TO_TEMP_OBJ" : {
```

File: core/src/main/resources/error/error-classes.json (hunk 1 above)
Comment (allisonwang-db) [reply core/src/main/resources/error/error-classes.json:182]: I think they are used interchangeably in Spark. Any preference here @srielau?

**Review by srielau — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 3]
```diff
@@ -2700,20 +2700,24 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
       name: TableIdentifier,
       nameParts: String): Throwable = {
     new AnalysisException(
-      errorClass = "_LEGACY_ERROR_TEMP_1283",
+      errorClass = "INVALID_TEMP_OBJ_REFERENCE",
       messageParameters = Map(
-        "name" -> name.toString,
-        "nameParts" -> nameParts))
+        "obj" -> "view",
+        "objName" -> name.toString,
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:2706]: Could you quote the table id by using `toSQLId`:
```suggestion
        "objName" -> toSQLId(name.nameParts),```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 4]
```diff
@@ -2700,20 +2700,24 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
...
+        "tempObj" -> "view",
+        "tempObjName" -> nameParts))
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:2708]: The `quote()` method uses `quoteIfNeeded`, please, use `toSQLId` to quote the name (it always quotes ids).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 5]
```diff
@@ -2700,20 +2700,24 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
...
       name: TableIdentifier,
       funcName: String): Throwable = {
      new AnalysisException(
-      errorClass = "_LEGACY_ERROR_TEMP_1284",
+      errorClass = "INVALID_TEMP_OBJ_REFERENCE",
       messageParameters = Map(
-        "name" -> name.toString,
-        "funcName" -> funcName))
+        "obj" -> "view",
+        "objName" -> name.toString,
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:2718]:
```suggestion
        "objName" -> toSQLId(name.nameParts),
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 6]
```diff
@@ -2700,20 +2700,24 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
...
+        "tempObj" -> "function",
+        "tempObjName" -> funcName))
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:2720]: Please, quote function id using `toSQLId()`.

**Review by MaxGekk — CHANGES_REQUESTED** [review]
@allisonwang-db Please, update PR's title and description w/ new error class name.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
The kub test is not related to this PR, I think:
```
[info] - SPARK-38187: Run SparkPi Jobs with minCPU *** FAILED *** (3 minutes, 1 second)
[info]   The code passed to eventually never returned normally. Attempted 189 times over 3.0106043718 minutes. Last failure message: 0 did not equal 2. (VolcanoTestsSuite.scala:302)
```

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master/3.4.
Thank you, @allisonwang-db.

**Pull request closed without merging**